receiver/k8sobjectsreceiver/receiver.go

```diff
@@ -49,22 +56,70 @@ func newReceiver(params receiver.CreateSettings, config *Config, consumer consum
 		return nil, err
 	}
 
-	return &k8sobjectsreceiver{
-		client:   client,
-		setting:  params,
-		consumer: consumer,
-		objects:  config.Objects,
-		obsrecv:  obsrecv,
-		mu:       sync.Mutex{},
-	}, nil
+	objReceiver := &k8sobjectsreceiver{
+		client:         client,
+		setting:        params,
+		logger:         params.Logger,
+		consumer:       consumer,
+		objects:        config.Objects,
+		obsrecv:        obsrecv,
+		mu:             sync.Mutex{},
+		leaderElection: config.LeaderElection,
+	}
+
+	if config.LeaderElection.Enabled {
+		if objReceiver.leaderElection.LockName == "" {
+			return nil, errors.New("luckName must not be empty if LeaderElection enabled")
+		}
+
+		objReceiver.leaderElectionClient, err = config.getClient()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return objReceiver, nil
+}
+
+func (kr *k8sobjectsreceiver) startFunc(ctx context.Context) {
+	for _, object := range kr.objects {
+		kr.start(ctx, object)
+	}
+}
+
+func (kr *k8sobjectsreceiver) startInLeaderElectionMode(ctx context.Context) {
+	leaderLost := make(chan struct{})
+
+	kr.mu.Lock()
+	kr.stopperChanList = append(kr.stopperChanList, leaderLost)
+	kr.mu.Unlock()
+
+	lr, err := k8sconfig.NewLeaderElector(kr.leaderElection, kr.leaderElectionClient, func(_ context.Context) {
+		kr.logger.Info("this instance of the component was selected as the current leader",
+			zap.String("lock name", kr.leaderElection.LockName))
+
+		kr.startFunc(ctx)
+	},
+		func() {
+			kr.logger.Error("this instance of the component was previously the leader but was removed as such")
+			leaderLost <- struct{}{}
+		})
```

Comment:
I think the receiver should stop collecting object when it is not the leader any more, and go back into the _wait_ mode?

Reply:
You are right, but once a pod is running as a leader, it will always be the leader, unless it crashes, but when it crashes, other replicas will attempt to become the leader, so the new pod(crashed before) will become a backup node, naturally in wait mode.

Reply:
> once a pod is running as a leader, it will always be the leader, unless it crashes
I don't think we should rely on this assumption. The non-leader candidates will wait the `LeaseDuration` (default 15sec) after the last successful renewal by leader to attempt to acquire the lock. It is possible that the leader pod is unable to renew the lease object for a short duration of time due to issues like network partition/congestion, etc. Once the old leader is back (as in can connect to k8s api again), another candidate could very well have acquired leadership

Reply:
thx @newly12  @jinja2 , I will update the logical here